### PR TITLE
Don't explicitly add default include paths to Extensions

### LINKED ISSUE #9737: setupext should not explicitly add /usr/{,local/}include to the include path
### Bug report

**Bug summary**

Currently, on Linux, setupext.py explicitly adds /usr/{,local/}include to the search path (because it is listed in get_base_dirs).  This can be confirmed by checking e.g. the output of `python setup.py build`.
This makes it impossible to compile matplotlib using a non-system compiler (e.g., anaconda now provides gcc7.2, see https://www.anaconda.com/blog/developer-blog/utilizing-the-new-compilers-in-anaconda-distribution-5/) that comes with its own headers (in /path/to/env/lib/gcc/x86_64-conda_cos6-linux-gnu/7.2.0/include) which may be incompatible with the system headers.  Note that conda's gcc is configured to indeed look into its own headers directory instead of /usr/include.

In the specific case of gcc7.2.0 from anaconda, explicitly adding /usr/include causes it to use /usr/include/math.h instead of its own math.h; Ubuntu's /usr/include/math.h is incompatible because it tries to include bits/mathdef.h and this file is under /usr/lib/gcc/x86_64-linux-gnu/5/include, which is part of the default path of the *system* gcc but not the conda gcc.

We currently explicitly look up these default paths to check that the include files are actually present, as a fallback in case they exist but no pkg-config information is present.  I propose instead to require pkg-config information to be present (as a check that the package is indeed installed) and not try to manually look for the headers.

This would cause a build failure if there is some linux distro that includes the headers for our dependencies *without* including the pkg-config info, which I am happy to claim to be a bad idea.

This would also mean that if one is using a non-system compiler (e.g., a conda compiler), then one would also need to install freetype/libpng/zlib in a place this compiler knows about (e.g., via a conda package) rather than relying on the system version.  Here too I think this is a reasonable restriction (if you know to set up your own non-system compiler you should know how to set up the dependencies properly too...).

**Code for reproduction**

```
$ conda create -n test -c conda-forge numpy freetype && source activate test && conda install -yc anaconda gxx_linux-64
$ <from matplotlib source> python setup.py build
```

**Actual outcome**

```
/home/alee/miniconda3/envs/test/bin/x86_64-conda_cos6-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DFREETYPE_BUILD_TYPE=system -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_ft2font_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -D__STDC_FORMAT_MACROS=1 -I/home/alee/miniconda3/envs/test/lib/python3.6/site-packages/numpy/core/include -I/home/alee/miniconda3/envs/test/include/freetype2 -I/usr/local/include -I/usr/include -I. -I/home/alee/miniconda3/envs/test/include/python3.6m -c src/ft2font.cpp -o build/temp.linux-x86_64-3.6/src/ft2font.o
In file included from /home/alee/miniconda3/envs/test/include/python3.6m/pyport.h:194:0,
                 from /home/alee/miniconda3/envs/test/include/python3.6m/Python.h:50,
                 from src/mplutils.h:31,
                 from src/ft2font.cpp:9:
/usr/include/math.h:31:10: fatal error: bits/math-vector.h: No such file or directory
 #include <bits/math-vector.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
error: command '/home/alee/miniconda3/envs/test/bin/x86_64-conda_cos6-linux-gnu-cc' failed with exit status 1
```

**Expected outcome**

Successful compilation.  In the example above, this can be achieved by patching setupext's get_base_dirs to return an empty list -- this works because pkg-config info is available for all packages (so we don't do any manual header lookup ourselves), giving information to gcc about where to find the headers.

**Matplotlib version**

  * Operating system: Ubuntu
  * Matplotlib version: master
  * Matplotlib backend (`print(matplotlib.get_backend())`): N/A
  * Python version: 3.6
  * Jupyter version (if applicable): N/A
  * Other libraries: N/A

-----

xref https://github.com/matplotlib/matplotlib/issues/3821 as well (ultimately we won't have /usr/local appear explicitly anywhere anymore in setupext.py).

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

Closes #9737 and makes setupext.py 20% shorter.

Builds on top of #11964; once that PR is merged I'm happy to further split this one into smaller pieces to help review if needed.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
rebased now that #11964 is merged.

File: setupext.py [hunk 1]
```diff
@@ -353,52 +351,28 @@ def set_pkgconfig_path(self):
...
+        """Add parameters to the given *ext* for the given *package*."""
+        cmd = ([self.pkg_config, package] if self.pkg_config
+                      else alt_exec)
+        if cmd is not None:
             try:
-                output = subprocess.check_output(
-                    command, shell=True, stderr=subprocess.STDOUT)
-            except subprocess.CalledProcessError:
+                flags = shlex.split(subprocess.check_output(
+                    [*cmd, "--cflags", "--libs"], universal_newlines=True))
```
Comment (QuLogic) [inline setupext.py]: This is potentially a different encoding than the file system encoding used below.

File: setupext.py [hunk 2]
```diff
@@ -806,12 +780,7 @@ def add_flags(self, ext):
         else:
             pkg_config.setup_extension(
                 ext, 'freetype2',
-                default_include_dirs=[
-                    'include/freetype2', 'freetype2',
-                    'lib/freetype2/include',
-                    'lib/freetype2/include/freetype2'],
-                default_library_dirs=[
-                    'freetype2/lib'],
+                alt_exec=['freetype-config', '--cflags', '--libs'],
```
Comment (QuLogic) [inline setupext.py]: Isn't `--cflags --libs` redundant here?

File: setupext.py (hunk 1 above)
Comment (anntzer) [reply setupext.py]: good catch, changed to use os.fsdecode() with a comment for clarification.

File: setupext.py (hunk 2 above)
Comment (anntzer) [reply setupext.py]: good point, fixed.

File: setupext.py [hunk 3]
```diff
@@ -375,65 +273,49 @@ def set_pkgconfig_path(self):
...
+                    subprocess.check_call(
+                        [*cmd, f"--atleast-version={atleast_version}"])
+                # Use sys.getfilesystemencoding() to allow round-tripping
+                # when passed back to later subprocess calls; do not use
+                # locale.getpreferredencoding() which universal_newlines=True
+                # would do.
+                cflags = shlex.split(
+                    os.fsdecode(subprocess.check_output([*cmd, "--cflags"])))
+                libs = shlex.split(
+                    os.fsdecode(subprocess.check_output([*cmd, "--libs"])))
```
Comment (QuLogic) [inline setupext.py]: Is the previously added (but now removed again) comment about `--ldflags` no longer true?

Comment (anntzer) [reply setupext.py]: For the libpng-config case (so only if pkg-config is not present), that means that the cflags variable will actually also contain the output given by libpng-config --ldflags... which is not a problem (per the now removed comment).

**Comment by tacaswell** [comment]
attn @sandrotosi @cgohlke @msarahan  don't think this will cause you problems, but just flagging this a head of 3.1

**Pull request merged**